README
NAME
    String::Numeric - Determine whether a string represents a numeric value

SYNOPSIS
        $boolean = is_float($string);
        $boolean = is_decimal($string);
    
        $boolean = is_int($string);
        $boolean = is_int8($string);
        $boolean = is_int16($string);
        $boolean = is_int32($string);
        $boolean = is_int64($string);
        $boolean = is_int128($string);
    
        $boolean = is_uint($string);
        $boolean = is_uint8($string);
        $boolean = is_uint16($string);
        $boolean = is_uint32($string);
        $boolean = is_uint64($string);
        $boolean = is_uint128($string);

DESCRIPTION
    A numeric value contains an integer part that may be prefixed with an
    optional minus sign, which may be followed by a fractional part and/or
    an exponent part.

    See "String::Numeric ABNF" for specification and "COMPARISON" for a
    comparison with "Scalar::Util::looks_like_number()".

FUNCTIONS
  is_float
    Determine whether $string is a floating-point number of arbitrary
    precision.

    *Usage*

        $boolean = is_float($string);
        $boolean = is_float('-1');     # true
        $boolean = is_float('1');      # true
        $boolean = is_float('1.0');    # true
        $boolean = is_float('1.0e6');  # true
        $boolean = is_float('1e6');    # true
        $boolean = is_float(undef);    # false

    *Arguments*

    $string

    *Returns*

    $boolean

    *Note*

    This function is also available as "is_numeric()"

  is_decimal
    Determine whether $string is a decimal number of arbitrary precision.

    *Usage*

        $boolean = is_decimal($string);
        $boolean = is_decimal('-1');     # true
        $boolean = is_decimal('1');      # true
        $boolean = is_decimal('1.0');    # true
        $boolean = is_decimal('1.0e6');  # false
        $boolean = is_decimal('1e6');    # false
        $boolean = is_decimal(undef);    # false

    *Arguments*

    $string

    *Returns*

    $boolean

  is_int
    Determine whether $string is an integer of arbitrary precision.

    *Usage*

        $boolean = is_int($string);
        $boolean = is_int('1');      # true
        $boolean = is_int('-1');     # true
        $boolean = is_int('1.0');    # false
        $boolean = is_int(undef);    # false

    *Arguments*

    $string

    *Returns*

    $boolean

    *Note*

    This function is also available as "is_integer()"

  is_int8
    Determine whether $string is a 8-bit signed integer which can have any
    value in the range -128 to 127.

    *Usage*

        $boolean = is_int8($string);
        $boolean = is_int8('-128');   # true
        $boolean = is_int8('127');    # true
        $boolean = is_int8(undef);    # false

    *Arguments*

    $string

    *Returns*

    $boolean

  is_int16
    Determine whether $string is a 16-bit signed integer which can have any
    value in the range -32,768 to 32,767.

    *Usage*

        $boolean = is_int16($string);
        $boolean = is_int16('-32768');   # true
        $boolean = is_int16('32767');    # true
        $boolean = is_int16(undef);      # false

    *Arguments*

    $string

    *Returns*

    $boolean

  is_int32
    Determine whether $string is a 32-bit signed integer which can have any
    value in the range -2,147,483,648 to 2,147,483,647.

    *Usage*

        $boolean = is_int32($string);
        $boolean = is_int32('-2147483648');   # true
        $boolean = is_int32('2147483647');    # true
        $boolean = is_int32(undef);           # false

    *Arguments*

    $string

    *Returns*

    $boolean

  is_int64
    Determine whether $string is a 64-bit signed integer which can have any
    value in the range -9,223,372,036,854,775,808 to
    9,223,372,036,854,775,807.

    *Usage*

        $boolean = is_int64($string);
        $boolean = is_int64('-9223372036854775808');   # true
        $boolean = is_int64('9223372036854775807');    # true
        $boolean = is_int64(undef);                    # false

    *Arguments*

    $string

    *Returns*

    $boolean

  is_int128
    Determine whether $string is a 128-bit signed integer which can have any
    value in the range -170,141,183,460,469,231,731,687,303,715,884,105,728
    to 170,141,183,460,469,231,731,687,303,715,884,105,727.

    *Usage*

        $boolean = is_int128($string);
        $boolean = is_int128('-170141183460469231731687303715884105728');   # true
        $boolean = is_int128('170141183460469231731687303715884105727');    # true
        $boolean = is_int128(undef);                                        # false

    *Arguments*

    $string

    *Returns*

    $boolean

  is_uint
    Determine whether $string is an unsigned integer of arbitrary precision.

    *Usage*

        $boolean = is_uint($string);
        $boolean = is_uint('1');      # true
        $boolean = is_uint('-1');     # false
        $boolean = is_uint('1.0');    # false
        $boolean = is_uint(undef);    # false

    *Arguments*

    $string

    *Returns*

    $boolean

  is_uint8
    Determine whether $string is a 8-bit unsigned integer which can have any
    value in the range 0 to 255.

    *Usage*

        $boolean = is_uint8($string);
        $boolean = is_uint8('0');      # true
        $boolean = is_uint8('255');    # true
        $boolean = is_uint8(undef);    # false

    *Arguments*

    $string

    *Returns*

    $boolean

  is_uint16
    Determine whether $string is a 16-bit unsigned integer which can have
    any value in the range 0 to 65,535.

    *Usage*

        $boolean = is_uint16($string);
        $boolean = is_uint16('0');      # true
        $boolean = is_uint16('65535');  # true
        $boolean = is_uint16(undef);    # false

    *Arguments*

    $string

    *Returns*

    $boolean

  is_uint32
    Determine whether $string is a 32-bit unsigned integer which can have
    any value in the range 0 to 4,294,967,295.

    *Usage*

        $boolean = is_uint32($string);
        $boolean = is_uint32('0');           # true
        $boolean = is_uint32('4294967295');  # true
        $boolean = is_uint32(undef);         # false

    *Arguments*

    $string

    *Returns*

    $boolean

  is_uint64
    Determine whether $string is a 64-bit unsigned integer which can have
    any value in the range 0 to 18,446,744,073,709,551,615.

    *Usage*

        $boolean = is_uint64($string);
        $boolean = is_uint64('0');                     # true
        $boolean = is_uint64('18446744073709551615');  # true
        $boolean = is_uint64(undef);                   # false

    *Arguments*

    $string

    *Returns*

    $boolean

  is_uint128
    Determine whether $string is a 128-bit unsigned integer which can have
    any value in the range 0 to
    340,282,366,920,938,463,463,374,607,431,768,211,455.

    *Usage*

        $boolean = is_uint128($string);
        $boolean = is_uint128('0');                                        # true
        $boolean = is_uint128('340282366920938463463374607431768211455');  # true
        $boolean = is_uint128(undef);                                      # false

    *Arguments*

    $string

    *Returns*

    $boolean

LIMITATIONS
    "String::Numeric" supports numbers in decimal notation using Western
    Arabic numerals and decimal fractions seperated with a dot. Other
    notations or numeral systems are not supported.

COMPARISON
  Scalar::Util
    None of the following is considered valid by "String::Numeric".

    * Fractional number without integer part or fractional part
            looks_like_number("1.") # true
            looks_like_number(".1") # true

    * Plus prefix
            looks_like_number("+1") # true

    * Leading zeros
            looks_like_number("01") # true

    * Radix point other than "."
            looks_like_number("1,1") # true assuming locale is in effect and locale radix point is ','

    * Leading and/or trailing whitespace, such as HT, LF, FF, CR, and SP
            looks_like_number("\n1")   # true
            looks_like_number("1\n")   # true
            looks_like_number("\n1\n") # true

    * Numeric values that are not represented as a sequence of digits, such
    as "Inf", "Infinity", "NaN" and "0 but true"
            looks_like_number("Inf")        # true
            looks_like_number("Infinity")   # true
            looks_like_number("NaN")        # true
            looks_like_number("0 but true") # true

GRAMMAR
  String::Numeric ABNF
        ; String::Numeric ABNF Grammar
    
        is-float          = [ "-" ] FloatNumber
        is-decimal        = [ "-" ] DecimalNumber
        is-integer        = [ "-" ] DecimalInteger
        is-uint           = DecimalInteger
    
        FloatNumber       = DecimalNumber [ ExponentPart ]
        DecimalNumber     = DecimalInteger [ FractionalPart ]
        DecimalInteger    = "0" / ( NonZeroDigit *DecimalDigit )
    
        DecimalDigit      = %x30-39; 0-9
        NonZeroDigit      = %x31-39; 1-9
        FractionalPart    = "." 1*DecimalDigit
        ExponentIndicator = "e" / "E"
        ExponentPart      = ExponentIndicator [ "+" / "-" ] 1*DecimalDigit

  looks_like_number ABNF
        ; ABNF Grammar based on 5.10 Perl_grok_number()
    
        looks-like-number = Number / ZeroButTrue
    
        Number            = *WS [ Sign ] ( FloatNumber / ( Infinity / NaN ) ) *WS;
        FloatNumber       = DecimalNumber [ ExponentPart ]
        DecimalNumber     = DecimalInteger [ "." *DecimalDigit ] / "." 1*DecimalDigit
        DecimalInteger    = 1*DecimalDigit
    
        WS                = %x09 ; HT
                          / %x0A ; LF 
                          / %x0C ; FF
                          / %x0D ; CR
                          / %x20 ; SP
        DecimalDigit      = %x30-39; 0-9
        Sign              = "+" / "-"
        ExponentIndicator = "e" / "E"
        ExponentPart      = ExponentIndicator [ Sign ] 1*DecimalDigit
        Infinity          = "Inf" / "Infinity"
        NaN               = "NaN"
        ZeroButTrue       = %x30.20.62.75.74.20.74.72.75.65 ; 0 but true

EXPORTS
    None by default. Functions and constants can either be imported
    individually or in sets grouped by tag names. The tag names are:

    ":all" exports all functions.
    ":is_int" exports all "is_int" functions.
    ":is_uint" exports all "is_uint" functions.

DIAGNOSTICS
    (F) Usage: %s(string)
        Subroutine %s invoked with wrong number of arguments.

ENVIRONMENT
    Set the environment variable "STRING_NUMERIC_PP" to a true value before
    loading this package to disable usage of XS implementation.

PREREQUISITES
  Run-Time
    perl 5.006 or greater.
    Carp, core module.
    Exporter, core module.

  Build-Time
    In addition to Run-Time:

    Test::More 0.47 or greater, core module since 5.6.2.
    Test::Exception.

SEE ALSO
    String::Numeric::XS

    String::Numeric::Regexp

    Data::Types

    Regexp::Common::number

    Scalar::Util

    Scalar::Util::Numeric

AUTHOR
    Christian Hansen "chansen@cpan.org"

COPYRIGHT AND LICENSE
    Copyright 2009-2010 Christian Hansen.

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
0.9_01 2010-02-06
    - Added String::Numeric::Regexp
    - Refactored String::Numeric::PP to use String::Numeric::Regexp
      - This gives the PP version a performance increase of ~ 40-80 percent.
      - XS version is ~ 400-600 percent faster than PP version and ~ 150-300
        percent faster than using regexp.

0.9 2009-11-20
    - Added more tests, 100% coverage.

0.2 2009-10-14
    - Improved documentation.

0.1 2009-10-12
    - Initial release.



lib/String/Numeric/PP.pm
package String::Numeric::PP;
use strict;
use warnings;

use Carp                    qw[croak];
use String::Numeric::Regexp qw[:all];

BEGIN {
    our $VERSION   = '0.9_01';
    our @EXPORT_OK = qw(
      is_numeric
      is_float
      is_decimal
      is_integer
      is_int
      is_int8
      is_int16
      is_int32
      is_int64
      is_int128
      is_uint
      is_uint8
      is_uint16
      is_uint32
      is_uint64
      is_uint128
    );

    require Exporter;
    *import = \&Exporter::import;
}

sub INT128_MIN  () { '170141183460469231731687303715884105728' }
sub INT128_MAX  () { '170141183460469231731687303715884105727' }
sub UINT128_MAX () { '340282366920938463463374607431768211455' }

*is_numeric = \&is_float;
*is_integer = \&is_int;

sub is_float {
    @_ == 1 || croak(q/Usage: is_float(string)/);
    return ( defined $_[0] && $_[0] =~ /\A$Float_re\z/o );
}

sub is_decimal {
    @_ == 1 || croak(q/Usage: is_decimal(string)/);
    return ( defined $_[0] && $_[0] =~ /\A$Decimal_re\z/o );
}

sub is_int {
    @_ == 1 || croak(q/Usage: is_int(string)/);
    return ( defined $_[0] && $_[0] =~ /\A$Int_re\z/o );
}

sub is_int8 {
    @_ == 1 || croak(q/Usage: is_int8(string)/);
    return ( defined $_[0] && $_[0] =~ /\A$Int8_re\z/o );
}

sub is_int16 {
    @_ == 1 || croak(q/Usage: is_int16(string)/);
    return ( defined $_[0] && $_[0] =~ /\A$Int16_re\z/o );
}

sub is_int32 {
    @_ == 1 || croak(q/Usage: is_int32(string)/);
    return ( defined $_[0] && $_[0] =~ /\A$Int32_re\z/o );
}

sub is_int64 {
    @_ == 1 || croak(q/Usage: is_int64(string)/);
    local $_ = $_[0];
    return ( defined $_[0] && $_[0] =~ /\A$Int64_re\z/o );
}

sub is_int128 {
    @_ == 1 || croak(q/Usage: is_int128(string)/);
    local $_ = $_[0];
    return ( defined
          && /\A(-?)(0|[1-9][0-9]{0,38})\z/
          && ( length $2 < 39 || ( $1 ? $2 le INT128_MIN : $2 le INT128_MAX ) ) );
}

sub is_uint {
    @_ == 1 || croak(q/Usage: is_uint(string)/);
    return ( defined $_[0] && $_[0] =~ /\A$UInt_re\z/o );
}

sub is_uint8 {
    @_ == 1 || croak(q/Usage: is_uint8(string)/);
    return ( defined $_[0] && $_[0] =~ /\A$UInt8_re\z/o );
}

sub is_uint16 {
    @_ == 1 || croak(q/Usage: is_uint16(string)/);
    return ( defined $_[0] && $_[0] =~ /\A$UInt16_re\z/o );
}

sub is_uint32 {
    @_ == 1 || croak(q/Usage: is_uint32(string)/);
    return ( defined $_[0] && $_[0] =~ /\A$UInt32_re\z/o );
}

sub is_uint64 {
    @_ == 1 || croak(q/Usage: is_uint64(string)/);
    return ( defined $_[0] && $_[0] =~ /\A$UInt64_re\z/o );
}

sub is_uint128 {
    @_ == 1 || croak(q/Usage: is_uint128(string)/);
    local $_ = $_[0];
    return ( defined
          && /\A(0|[1-9][0-9]{0,38})\z/
          && ( length $1 < 39 || $1 le UINT128_MAX ) );
}

1;



lib/String/Numeric/Regexp.pm
package String::Numeric::Regexp;
use strict;
use warnings;

BEGIN {
    our $VERSION   = '0.9_01';
    our @EXPORT_OK = qw(
        $Float_re
        $Decimal_re
        $Numeric_re
        $Int_re
        $Int8_re
        $Int16_re
        $Int32_re
        $Int64_re
        $UInt_re
        $UInt8_re
        $UInt16_re
        $UInt32_re
        $UInt64_re
    );

    our %EXPORT_TAGS = ( 
        all  => [ @EXPORT_OK ], 
        int  => [ grep { /^\$Int/ }  @EXPORT_OK ],
        uint => [ grep { /^\$UInt/ } @EXPORT_OK ],
    );

    require Exporter;
    *import = \&Exporter::import;
}

our $Float_re   = qr<-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?>;
our $Decimal_re = qr<-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?>;
our $Numeric_re = $Float_re;

our $UInt_re = qr<0|[1-9][0-9]*>;

# 0 to 127
our $UInt7_re = qr<
       0
  | [1-9][0-9]{0,1}
  |    1 (?:   [0-1][0-9]
  |        | 2 [0-7])
>x;

# 0 to 255
our $UInt8_re = qr<
       0
  |    1 [0-9]{0,2}
  | [2-9][0-9]{0,1}
  |    2 (?:   [0-4][0-9]
           | 5 [0-5])
>x;

# 0 to 32,767
our $UInt15_re = qr<
       0
  | [1-2][0-9]{0,4}
  | [3-9][0-9]{0,3}
  |    3 (?:       [0-1][0-9]{3}
           | 2 (?: [0-6][0-9]{2}
           | 7 (?: [0-5][0-9]
           | 6 [0-7])))
>x;

# 0 to 65,535
our $UInt16_re = qr<
       0
  | [1-5][0-9]{0,4}
  | [6-9][0-9]{0,3}
  |    6 (?:       [0-4][0-9]{3}
           | 5 (?: [0-4][0-9]{2}
           | 5 (?: [0-2][0-9]
           | 3 [0-5])))
>x;

# 0 to 2,147,483,647
our $UInt31_re = qr<
       0
  |    1 [0-9]{0,9}
  | [2-9][0-9]{0,8}
  |    2 (?: 0          [0-9]{8}
           | 1 (?: [0-3][0-9]{7}
           | 4 (?: [0-6][0-9]{6}
           | 7 (?: [0-3][0-9]{5}
           | 4 (?: [0-7][0-9]{4}
           | 8 (?: [0-2][0-9]{3}
           | 3 (?: [0-5][0-9]{2}
           | 6 (?: [0-3][0-9]
           | 4 [0-7]))))))))
>x;

# 0 to 4,294,967,295
our $UInt32_re = qr<
       0
  | [1-3][0-9]{0,9}
  | [4-9][0-9]{0,8}
  |    4 (?:       [0-1][0-9]{8}
           | 2 (?: [0-8][0-9]{7}
           | 9 (?: [0-3][0-9]{6}
           | 4 (?: [0-8][0-9]{5}
           | 9 (?: [0-5][0-9]{4}
           | 6 (?: [0-6][0-9]{3}
           | 7 (?: [0-1][0-9]{2}
           | 2 (?: [0-8][0-9]
           | 9 [0-5]))))))))
>x;

# 0 to 9,223,372,036,854,775,807
our $UInt63_re = qr<
       0
  | [1-8][0-9]{0,18}
  |    9 [0-9]{0,17}
  |    9 (?:        [0-1][0-9]{17}
           | 2  (?: [0-1][0-9]{16}
           | 2  (?: [0-2][0-9]{15}
           | 3  (?: [0-2][0-9]{14}
           | 3  (?: [0-6][0-9]{13}
           | 7  (?: [0-1][0-9]{12}
           | 20 (?: [0-2][0-9]{10}
           | 3  (?: [0-5][0-9]{9}
           | 6  (?: [0-7][0-9]{8}
           | 8  (?: [0-4][0-9]{7}
           | 5  (?: [0-3][0-9]{6}
           | 4  (?: [0-6][0-9]{5}
           | 7  (?: [0-6][0-9]{4}
           | 7  (?: [0-4][0-9]{3}
           | 5  (?: [0-7][0-9]{2}
           | 80 [0-7])))))))))))))))
>x;

# 0 to 18,446,744,073,709,551,615
our $UInt64_re = qr<
       0
  | [1-9][0-9]{0,18}
  |    1 (?:        [0-7][0-9]{18}
           | 8  (?: [0-3][0-9]{17}
           | 4  (?: [0-3][0-9]{16}
           | 4  (?: [0-5][0-9]{15}
           | 6  (?: [0-6][0-9]{14}
           | 7  (?: [0-3][0-9]{13}
           | 4  (?: [0-3][0-9]{12}
           | 40 (?: [0-6][0-9]{10}
           | 7  (?: [0-2][0-9]{9}
           | 3  (?: [0-6][0-9]{8}
           | 70 (?: [0-8][0-9]{6}
           | 9  (?: [0-4][0-9]{5}
           | 5  (?: [0-4][0-9]{4}
           | 5  (?:    0 [0-9]{3}
           | 1  (?: [0-5][0-9]{2}
           | 6  (?:    0 [0-9]
           | 1 [0-5]))))))))))))))))
>x;

our $Int_re = qr<-?(?:0|[1-9][0-9]*)>;

# −128 to 127
our $Int8_re = qr<(?: - (?: $UInt7_re | 128 )) | $UInt7_re>x;

# −32,768 to 32,767
our $Int16_re = qr<(?: - (?: $UInt15_re | 32768 )) | $UInt15_re>x;

# −2,147,483,648 to 2,147,483,647
our $Int32_re = qr<(?: - (?: $UInt31_re | 2147483648 )) | $UInt31_re>x;

# −9,223,372,036,854,775,808 to +9,223,372,036,854,775,807
our $Int64_re = qr<(?: - (?: $UInt63_re | 9223372036854775808 )) | $UInt63_re>x;

1;

__END__

=head1 NAME

String::Numeric::Regexp - Regular expressions for matching numeric values

=head1 SYNOPSIS

    if (/\A$UInt32\z/o) {
        print "ok"
    }

=head1 DESCRIPTION

Provides regexps to match numeric values. These are used internally by
C<String::Numeric::PP>. Please see L<String::Numeric> for expected matches.

=head1 REGEXPS

=head2 C<$Float_re>

=head2 C<$Decimal_re>

=head2 C<$Numeric_re>

Alias to C<$Float_re>.

=head2 C<$Int_re>

=head2 C<$Int8_re>

=head2 C<$Int16_re>

=head2 C<$Int32_re>

=head2 C<$Int64_re>

=head2 C<$UInt_re>

=head2 C<$UInt8_re>

=head2 C<$UInt16_re>

=head2 C<$UInt32_re>

=head2 C<$UInt64_re>

=head1 EXPORTS

None by default. Regexps can either be imported individually or
in sets grouped by tag names. The tag names are:

=over 4

=item C<:all> exports all regexps.

=item C<:int> exports all C<$Int> regexps.

=item C<:uint> exports all C<$UInt> regexps.

=back

=head1 SEE ALSO

L<String::Numeric>

L<Regexp::Common::number>

=head1 AUTHOR

Christian Hansen C<chansen@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright 2009-2010 Christian Hansen.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/String/Numeric.pm
package String::Numeric;
use strict;
use warnings;

BEGIN {
    our $VERSION   = '0.9_01';
    our @EXPORT_OK = qw(
        is_numeric
        is_float
        is_decimal
        is_integer
        is_int
        is_int8
        is_int16
        is_int32
        is_int64
        is_int128
        is_uint
        is_uint8
        is_uint16
        is_uint32
        is_uint64
        is_uint128
    );

    our %EXPORT_TAGS = ( 
        all  => [ @EXPORT_OK ], 
        int  => [ grep { /^is_int/ }  @EXPORT_OK ],
        uint => [ grep { /^is_uint/ } @EXPORT_OK ],
    );

    if ( !$ENV{STRING_NUMERIC_PP} ) {
        eval {
            require String::Numeric::XS;
        };
    }

    if ( $ENV{STRING_NUMERIC_PP} || $@ ) {
        require String::Numeric::PP;
        String::Numeric::PP->import(@EXPORT_OK);
    }
    else {
        String::Numeric::XS->import(@EXPORT_OK);
    }

    require Exporter;
    *import = \&Exporter::import;
}

1;

__END__

=head1 NAME

String::Numeric - Determine whether a string represents a numeric value

=head1 SYNOPSIS

    $boolean = is_float($string);
    $boolean = is_decimal($string);
    
    $boolean = is_int($string);
    $boolean = is_int8($string);
    $boolean = is_int16($string);
    $boolean = is_int32($string);
    $boolean = is_int64($string);
    $boolean = is_int128($string);
    
    $boolean = is_uint($string);
    $boolean = is_uint8($string);
    $boolean = is_uint16($string);
    $boolean = is_uint32($string);
    $boolean = is_uint64($string);
    $boolean = is_uint128($string);

=head1 DESCRIPTION

A numeric value contains an integer part that may be prefixed with an 
optional minus sign, which may be followed by a fractional part and/or an 
exponent part.

See L</String::Numeric ABNF> for specification and L</COMPARISON> for a comparison with C<Scalar::Util::looks_like_number()>.

=head1 FUNCTIONS

=head2 is_float

Determine whether C<$string> is a floating-point number of arbitrary precision.

I<Usage>

    $boolean = is_float($string);
    $boolean = is_float('-1');     # true
    $boolean = is_float('1');      # true
    $boolean = is_float('1.0');    # true
    $boolean = is_float('1.0e6');  # true
    $boolean = is_float('1e6');    # true
    $boolean = is_float(undef);    # false

I<Arguments>

=over 4

=item C<$string>

=back

I<Returns>

=over 4

=item C<$boolean>

=back

I<Note>

=over 4

=item This function is also available as C<is_numeric()>

=back

=head2 is_decimal

Determine whether C<$string> is a decimal number of arbitrary precision.

I<Usage>

    $boolean = is_decimal($string);
    $boolean = is_decimal('-1');     # true
    $boolean = is_decimal('1');      # true
    $boolean = is_decimal('1.0');    # true
    $boolean = is_decimal('1.0e6');  # false
    $boolean = is_decimal('1e6');    # false
    $boolean = is_decimal(undef);    # false

I<Arguments>

=over 4

=item C<$string>

=back

I<Returns>

=over 4

=item C<$boolean>

=back

=head2 is_int

Determine whether C<$string> is an integer of arbitrary precision.

I<Usage>

    $boolean = is_int($string);
    $boolean = is_int('1');      # true
    $boolean = is_int('-1');     # true
    $boolean = is_int('1.0');    # false
    $boolean = is_int(undef);    # false

I<Arguments>

=over 4

=item C<$string>

=back

I<Returns>

=over 4

=item C<$boolean>

=back

I<Note>

=over 4

=item This function is also available as C<is_integer()>

=back

=head2 is_int8

Determine whether C<$string> is a 8-bit signed integer which can have any 
value in the range -128 to 127.

I<Usage>

    $boolean = is_int8($string);
    $boolean = is_int8('-128');   # true
    $boolean = is_int8('127');    # true
    $boolean = is_int8(undef);    # false

I<Arguments>

=over 4

=item C<$string>

=back

I<Returns>

=over 4

=item C<$boolean>

=back

=head2 is_int16

Determine whether C<$string> is a 16-bit signed integer which can have any 
value in the range -32,768 to 32,767.

I<Usage>

    $boolean = is_int16($string);
    $boolean = is_int16('-32768');   # true
    $boolean = is_int16('32767');    # true
    $boolean = is_int16(undef);      # false

I<Arguments>

=over 4

=item C<$string>

=back

I<Returns>

=over 4

=item C<$boolean>

=back

=head2 is_int32

Determine whether C<$string> is a 32-bit signed integer which can have any 
value in the range -2,147,483,648 to 2,147,483,647.

I<Usage>

    $boolean = is_int32($string);
    $boolean = is_int32('-2147483648');   # true
    $boolean = is_int32('2147483647');    # true
    $boolean = is_int32(undef);           # false

I<Arguments>

=over 4

=item C<$string>

=back

I<Returns>

=over 4

=item C<$boolean>

=back

=head2 is_int64

Determine whether C<$string> is a 64-bit signed integer which can have any 
value in the range -9,223,372,036,854,775,808 to 9,223,372,036,854,775,807.

I<Usage>

    $boolean = is_int64($string);
    $boolean = is_int64('-9223372036854775808');   # true
    $boolean = is_int64('9223372036854775807');    # true
    $boolean = is_int64(undef);                    # false

I<Arguments>

=over 4

=item C<$string>

=back

I<Returns>

=over 4

=item C<$boolean>

=back

=head2 is_int128

Determine whether C<$string> is a 128-bit signed integer which can have any 
value in the range -170,141,183,460,469,231,731,687,303,715,884,105,728 to 
170,141,183,460,469,231,731,687,303,715,884,105,727.

I<Usage>

    $boolean = is_int128($string);
    $boolean = is_int128('-170141183460469231731687303715884105728');   # true
    $boolean = is_int128('170141183460469231731687303715884105727');    # true
    $boolean = is_int128(undef);                                        # false

I<Arguments>

=over 4

=item C<$string>

=back

I<Returns>

=over 4

=item C<$boolean>

=back

=head2 is_uint

Determine whether C<$string> is an unsigned integer of arbitrary precision.

I<Usage>

    $boolean = is_uint($string);
    $boolean = is_uint('1');      # true
    $boolean = is_uint('-1');     # false
    $boolean = is_uint('1.0');    # false
    $boolean = is_uint(undef);    # false

I<Arguments>

=over 4

=item C<$string>

=back

I<Returns>

=over 4

=item C<$boolean>

=back

=head2 is_uint8

Determine whether C<$string> is a 8-bit unsigned integer which can have any 
value in the range 0 to 255.

I<Usage>

    $boolean = is_uint8($string);
    $boolean = is_uint8('0');      # true
    $boolean = is_uint8('255');    # true
    $boolean = is_uint8(undef);    # false

I<Arguments>

=over 4

=item C<$string>

=back

I<Returns>

=over 4

=item C<$boolean>

=back

=head2 is_uint16

Determine whether C<$string> is a 16-bit unsigned integer which can have any 
value in the range 0 to 65,535.

I<Usage>

    $boolean = is_uint16($string);
    $boolean = is_uint16('0');      # true
    $boolean = is_uint16('65535');  # true
    $boolean = is_uint16(undef);    # false

I<Arguments>

=over 4

=item C<$string>

=back

I<Returns>

=over 4

=item C<$boolean>

=back

=head2 is_uint32

Determine whether C<$string> is a 32-bit unsigned integer which can have any 
value in the range 0 to 4,294,967,295.

I<Usage>

    $boolean = is_uint32($string);
    $boolean = is_uint32('0');           # true
    $boolean = is_uint32('4294967295');  # true
    $boolean = is_uint32(undef);         # false

I<Arguments>

=over 4

=item C<$string>

=back

I<Returns>

=over 4

=item C<$boolean>

=back

=head2 is_uint64

Determine whether C<$string> is a 64-bit unsigned integer which can have any 
value in the range 0 to 18,446,744,073,709,551,615.

I<Usage>

    $boolean = is_uint64($string);
    $boolean = is_uint64('0');                     # true
    $boolean = is_uint64('18446744073709551615');  # true
    $boolean = is_uint64(undef);                   # false

I<Arguments>

=over 4

=item C<$string>

=back

I<Returns>

=over 4

=item C<$boolean>

=back

=head2 is_uint128

Determine whether C<$string> is a 128-bit unsigned integer which can have any 
value in the range 0 to 340,282,366,920,938,463,463,374,607,431,768,211,455.

I<Usage>

    $boolean = is_uint128($string);
    $boolean = is_uint128('0');                                        # true
    $boolean = is_uint128('340282366920938463463374607431768211455');  # true
    $boolean = is_uint128(undef);                                      # false

I<Arguments>

=over 4

=item C<$string>

=back

I<Returns>

=over 4

=item C<$boolean>

=back

=head1 LIMITATIONS

C<String::Numeric> supports numbers in decimal notation using Western Arabic
numerals and decimal fractions seperated with a dot. Other notations or numeral
systems are not supported.

=head1 COMPARISON

=head2 Scalar::Util

B<None> of the following is considered valid by C<String::Numeric>.

=over 4

=item * Fractional number without integer part or fractional part

    looks_like_number("1.") # true
    looks_like_number(".1") # true

=item * Plus prefix

    looks_like_number("+1") # true

=item * Leading zeros

    looks_like_number("01") # true

=item * Radix point other than C<.>

    looks_like_number("1,1") # true assuming locale is in effect and locale radix point is ','

=item * Leading and/or trailing whitespace, such as HT, LF, FF, CR, and SP

    looks_like_number("\n1")   # true
    looks_like_number("1\n")   # true
    looks_like_number("\n1\n") # true

=item * Numeric values that are not represented as a sequence of digits, such as 
C<Inf>, C<Infinity>, C<NaN> and C<0 but true>

    looks_like_number("Inf")        # true
    looks_like_number("Infinity")   # true
    looks_like_number("NaN")        # true
    looks_like_number("0 but true") # true

=back

=head1 GRAMMAR

=head2 String::Numeric ABNF

    ; String::Numeric ABNF Grammar
    
    is-float          = [ "-" ] FloatNumber
    is-decimal        = [ "-" ] DecimalNumber
    is-integer        = [ "-" ] DecimalInteger
    is-uint           = DecimalInteger
    
    FloatNumber       = DecimalNumber [ ExponentPart ]
    DecimalNumber     = DecimalInteger [ FractionalPart ]
    DecimalInteger    = "0" / ( NonZeroDigit *DecimalDigit )
    
    DecimalDigit      = %x30-39; 0-9
    NonZeroDigit      = %x31-39; 1-9
    FractionalPart    = "." 1*DecimalDigit
    ExponentIndicator = "e" / "E"
    ExponentPart      = ExponentIndicator [ "+" / "-" ] 1*DecimalDigit

=head2 looks_like_number ABNF

    ; ABNF Grammar based on 5.10 Perl_grok_number()
    
    looks-like-number = Number / ZeroButTrue
    
    Number            = *WS [ Sign ] ( FloatNumber / ( Infinity / NaN ) ) *WS;
    FloatNumber       = DecimalNumber [ ExponentPart ]
    DecimalNumber     = DecimalInteger [ "." *DecimalDigit ] / "." 1*DecimalDigit
    DecimalInteger    = 1*DecimalDigit
    
    WS                = %x09 ; HT
                      / %x0A ; LF 
                      / %x0C ; FF
                      / %x0D ; CR
                      / %x20 ; SP
    DecimalDigit      = %x30-39; 0-9
    Sign              = "+" / "-"
    ExponentIndicator = "e" / "E"
    ExponentPart      = ExponentIndicator [ Sign ] 1*DecimalDigit
    Infinity          = "Inf" / "Infinity"
    NaN               = "NaN"
    ZeroButTrue       = %x30.20.62.75.74.20.74.72.75.65 ; 0 but true

=head1 EXPORTS

None by default. Functions and constants can either be imported individually or
in sets grouped by tag names. The tag names are:

=over 4

=item C<:all> exports all functions.

=item C<:is_int> exports all C<is_int> functions.

=item C<:is_uint> exports all C<is_uint> functions.

=back

=head1 DIAGNOSTICS

=over 4

=item B<(F)> Usage: %s(string)

Subroutine %s invoked with wrong number of arguments.

=back

=head1 ENVIRONMENT

Set the environment variable C<STRING_NUMERIC_PP> to a true value before loading this package to disable usage of XS implementation.

=head1 PREREQUISITES

=head2 Run-Time

=over 4

=item L<perl> 5.006 or greater.

=item L<Carp>, core module.

=item L<Exporter>, core module.

=back

=head2 Build-Time

In addition to Run-Time:

=over 4

=item L<Test::More> 0.47 or greater, core module since 5.6.2.

=item L<Test::Exception>.

=back

=head1 SEE ALSO

L<String::Numeric::XS>

L<String::Numeric::Regexp>

L<Data::Types>

L<Regexp::Common::number>

L<Scalar::Util>

L<Scalar::Util::Numeric>

=head1 AUTHOR

Christian Hansen C<chansen@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright 2009-2010 Christian Hansen.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

